src/browser_action/on-each-interaction.js

```diff
@@ -11,43 +11,139 @@
+    if (loaf.startTime + loaf.duration < start) continue;
+
+    // If the LoAF starts after the given end time, ignore it and all
+    // subsequent pending LoAFs (because they're in time order).
+    if (loaf.startTime > end) break;
+
+    // Still here? If so this LoAF intersects with the interaction.
+    intersectingLoAFs.push(loaf);
+  }
+  return intersectingLoAFs;
+};
+
+const loafObserver = new PerformanceObserver((list) => {
+
+  for (const entry of list.getEntries()) { recentLoAFs.push(entry) };
+  // We report interactions immediately, so don't need to keep many LoAFs around.
+  // Let's keep the last 5.
+  recentLoAFs = recentLoAFs.slice(-5);
+
+});
+loafObserver.observe({
+  type: 'long-animation-frame',
+  buffered: true,
+});
+
 /**
+ * We emulate an INP entry with similar logic to web-vitals.js
+ * But we have it easier as will emit it immediately and don't need the p98 stuff
+ * So can strip a lot of that logic out.
  * @param {Function} callback
  */
 export function onEachInteraction(callback) {
+
   const valueToRating = (score) => score <= 200 ? 'good' : score <= 500 ? 'needs-improvement' : 'poor';
 
   const observer = new PerformanceObserver((list) => {
-    const interactions = {};
 
-    for (const entry of list.getEntries().filter((entry) => entry.interactionId)) {
-      interactions[entry.interactionId] = interactions[entry.interactionId] || [];
-      interactions[entry.interactionId].push(entry);
+    const allEvents = list.getEntries();
+
+    // filter just for INP-eligible events (those with an interactionId)
+    const interactions = allEvents.filter((entry) => entry.interactionId)
+    // If none then can just end
+    if (interactions.length == 0 ) return;
+
+    const longestEntry = interactions.reduce((prev, curr) => prev.duration >= curr.duration ? prev : curr);
+    const largestDuration = longestEntry.duration;
+    const largestRenderTime = longestEntry.startTime + longestEntry.duration;
+
+    const longestFrameEntries = [];
+    const group = {
+      startTime: null,
+      processingStart: null,
+      processingEnd: null
     }
 
-    // Will report as a single interaction even if parts are in separate frames.
-    // Consider splitting by animation frame.
-    for (const interaction of Object.values(interactions)) {
-      const entry = interaction.reduce((prev, curr) => prev.duration >= curr.duration ? prev : curr);
-      const value = entry.duration;
-
-      callback({
-        attribution: {
-          eventEntry: entry,
-          eventTime: entry.startTime,
-          eventType: entry.name,
-        },
-        entries: interaction,
-        name: 'Interaction',
-        rating: valueToRating(value),
-        value,
-      });
+    // Go though the list of all events (not just those with interactionIds)
+    // do get details of what was in the longest frame and the earliest and latest event timestamps
+    for (const entry of allEvents) {
```

Comment:
I'm not sure if it's guaranteed that all events that occur within the same frame will get dispatched as part of the same PO callback, @mmocny do you know?

The current logic in the web-vitals library assumes that's *not* necessarily the case, which is why it has the `pendingEntriesGroupMap` object.

Reply:
Since we want to report as the events are received, I'm not sure it's wanted to have `pendingEntriesGroupMap` and the complexity it brings (though I agree it's important for the actual, official INP).

So I'm comfortable with the slight risk here to allow early reporting and keep the implementation simpler.

Reply:
As per the other comment for typing this is definitely a risk, but also not convinced that's a huge issue (though I can imagine this upsets you to no end that it's technically incorrect!).

As I said the aim of this is to be able to quickly identify slow **discrete** interactions and IMHO typing isn't discrete and so reporting that only once isn't necessarily a bad thing.

Also [from the blog post where we introduced this](https://web.dev/articles/debug-cwvs-with-web-vitals-extension#interactions) there were three main reason:

> 1. INP can be slow due to previous interactions blocking the main thread, and thus causing a high input delay. For this reason, we list all interactions in a similar format to the INP logging

This is maybe less relevant now the INP event will have LoAFs to help identify this.

> 2. This allows you to "live trace" a website by interacting with it and seeing in the console which interactions, in which combinations, are likely to cause an INP problem.

This is still it's most useful feature. Maybe the three phase breakdown and LoAF inclusions aren't strictly necessary for that, but wanted to make it similar to the INP entry rather than just the 

>3. This also allows you to identify multiple slow interactions, rather than just the largest INP interaction to help you avoid the feeling of chasing your tail when improving your responsiveness.

Again to me what's most important is to identify long discrete interactions. Not being 100% accurate for the typing case or other long delayed interactions wasn't my primary goal as think a lot of the value has been lost then anyway.

Saying all that, I'll have another look and see how easy it is to report these properly. It shouldn't be too much complication in which case it's worth it for the accuracy stakes even if I don't think it will be massively useful in those cases.

> TBH, I think it would be better to rework the web-vitals INP attribution code so it could be shared with the extension than it would be to have the extension use a different definition of the INP breakdown.

That would be ideal! But not sure how easy that is without making the library worse for it—which we definitely don't want.

Reply:
> As per the other comment for typing this is definitely a risk, but also not convinced that's a huge issue

It's not just from typing, I frequently see cases where entries are dispatched in separate PO callback even when just clicking.

> (though I can imagine this upsets you to no end that it's technically incorrect!).

We've seen examples where external folks look at the code in this extension as a reference to build their features (e.. RUMvision), so it's important that it's correct.

> As I said the aim of this is to be able to quickly identify slow discrete interactions and IMHO typing isn't discrete and so reporting that only once isn't necessarily a bad thing.

Nit: we've always defined key presses as "discrete" interactions ([see the FID doc](https://web.dev/articles/fid#:~:text=FID%20is%20a%20metric%20that%20measures%20a%20page%27s%20responsiveness%20during%20load.%20As%20such%2C%20it%20only%20focuses%20on%20input%20events%20from%20discrete%20actions%20like%20clicks%2C%20taps%2C%20and%20key%20presses.)), as opposed to "continuous" interactions liike scroll or pointermove.

> > TBH, I think it would be better to rework the web-vitals INP attribution code so it could be shared with the extension than it would be to have the extension use a different definition of the INP breakdown.
> 
> That would be ideal! But not sure how easy that is without making the library worse for it—which we definitely don't want.

I don't think it would be too hard. I have some ideas that I don't think it would impact the web-vitals library at all (can share offline).

Reply:
OK as discussed the complexity to implement this properly would mean copying a LOT of code from web-vitals or making that exportable from web-vitals and using a bundler for this project (plus then they will be part of the public API for web-vitals).

Given that we don't think it's worth implementing the phased breakdown for the Interactions. In most part they will be used to identify slow events so the breakdowns may not even be used. So while having the break downs when they come for free (as they do with INP since web-vitals.js does the hard work) is fine, doing all that work for Interactions—and having to maintain that in future—when they are not for free means this is less desirable.

We will still add LoAF as that is a lot easier.